### PR TITLE
Add debug string to raylet

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds `DebugString()` calls to most raylet components. The idea is that this will log the raylet state for debugging at with a configurable frequency.

Monitor your raylet with `watch cat /tmp/ray/debug_state.txt`

- [x] TODO: make the logging frequency configurable via flag
- [x] TODO: can we always log to a file or trigger this in crash handlers?

cc @pcmoritz 

```
I1114 00:24:00.443905 22101 node_manager.cc:281] NodeManager:
LocalResources: 
- total: {num_cpus,1.000000},{num_gpus,0.000000},{CPU,4.000000},{GPU,0.000000}
- avail: {num_cpus,1.000000},{num_gpus,0.000000},{CPU,4.000000},{GPU,0.000000}
ClusterResources:
86918ef8a1e188ec2c1acb07b49c9520b0f3cc85: 
- total: {GPU,0.000000},{CPU,4.000000},{num_gpus,0.000000},{num_cpus,1.000000}
- avail: {GPU,0.000000},{CPU,2.000000},{num_gpus,0.000000},{num_cpus,1.000000}
7a4a0a6ddad0367c52ac83c5b496158ea966b147: 
- total: {num_cpus,1.000000},{num_gpus,0.000000},{CPU,4.000000},{GPU,0.000000}
- avail: {num_cpus,1.000000},{num_gpus,0.000000},{CPU,3.000000},{GPU,0.000000}
ObjectManager:
- num local objects: 27464
- num active wait requests: 1
- num unfulfilled push requests: 0
- num pull requests: 1
- num buffered profile events: 39
ObjectDirectory:
- num listeners: 1
- num eviction entries: 27464
ObjectStoreNotificationManager:
- num adds processed: 27464
- num removes processed: 0
BufferPool:
- get buffer state map size: 1
- create buffer state map size: 0
ConnectionPool:
- num message send connections: 1
- num transfer send connections: 1
- num avail message send connections: 1
- num avail transfer send connections: 1
- num message receive connections: 1
- num transfer receive connections: 1
AsyncGcsClient:
- TaskTable: Num lookups: 0, Num adds: 26381
- ActorTable: Num lookups: 2, Num appends: 2
- TaskReconstructionLog: Num lookups: 0, Num appends: 0
- TaskLeaseTable: Num lookups: 0, Num adds: 26397
- HeartbeatTable: Num lookups: 0, Num adds: 2951
- ErrorTable: Num lookups: 0, Num appends: 0
- ProfileTable: Num lookups: 0, Num appends: 3074
- ClientTable: Num lookups: 0, Num appends: 1, Cache size: 2, Num removed: 0
- DriverTable: Num lookups: 0, Num appends: 0
WorkerPool:
- num workers: 4
- num drivers: 1
SchedulingQueue:
- num placeable tasks: 0
- num waiting tasks: 7
- num ready tasks: 0
- num running tasks: 2
- num infeasible tasks: 0
- num methods waiting for actor creation: 0
ReconstructionPolicy:
- num reconstructing: 1
TaskDependencyManager:
- task dep map size: 8
- task req map size: 12
- req objects map size: 1
- local objects map size: 53843
- pending tasks map size: 9
LineageCache:
- committed tasks: 0
- child map size: 8741
- num subscribed tasks: 0
- lineage size: 7
RemoteConnections:
86918ef8a1e188ec2c1acb07b49c9520b0f3cc85: 
- bytes read: 0
- bytes written: 3920
- num async writes: 5
- num sync writes: 0
- writing: 0
- pending async bytes: 0
ActorRegistry:
bc5a9939adb6cedd194a7b54351ed59806999406: alive, num handles: 1
6ecf8cd36cd11cd995b55f202e46640bf12f9678: alive, num handles: 1
15e36d696088a3aa0ee21a4f10516275102a0c19: alive, num handles: 1
593246b5c9546d387fbfaaf2552833c372acbce2: alive, num handles: 1
DebugString() time ms: 0
```

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -1655,6 +1665,42 @@ void NodeManager::ForwardTask(const Task &task, const ClientID &node_id,
       });
 }
 
+void NodeManager::DumpDebugState() {
+  std::fstream fs;
+  // TODO(ekl) should we make this path configurable?
```
Comment (pcmoritz) [inline src/ray/raylet/node_manager.cc]: can we put this into the appropriate session subfolder?

**Comment by pcmoritz** [comment]
Thanks, this looks great!

Do you think it would be good to make it so this can be parsed with a yaml parser? It's already very close I think and adding a few underscores at the right places might be enough.

**Comment by ericl** [comment]
How do you get the session folder?

**Comment by pcmoritz** [comment]
https://github.com/ray-project/ray/blob/master/python/ray/tempfile_services.py#L68

**Comment by pcmoritz** [comment]
It needs to be passed through like this: https://github.com/ray-project/ray/blob/d4fad222e19b0cb39d66b418705b166415150ef5/python/ray/services.py#L957
and stored in the config, but that should be easy

**Comment by ericl** [comment]
Updated to put the file in the temp root.

Re: yaml, I don't think this is something we need to tie DebugString() to, if we need to get the info into structured formats we should do that explicitly.

File: src/ray/raylet/actor_registration.cc [hunk 2]
```diff
@@ -1,5 +1,6 @@
-#include "ray/raylet/actor_registration.h"
+#include <sstream>
 
+#include "ray/raylet/actor_registration.h"
```
Comment (robertnishihara) [inline src/ray/raylet/actor_registration.cc]: Should be

```cpp
#include "ray/raylet/actor_registration.h"

#include <sstream>

#include "ray/util/logging.h"
```

I think

File: src/ray/raylet/lineage_cache.cc [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#include <sstream>
```
Comment (robertnishihara) [inline src/ray/raylet/lineage_cache.cc]: Should be

```cpp
#include "lineage_cache.h"

#include <sstream>
```

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -1,3 +1,5 @@
+#include <fstream>
```
Comment (robertnishihara) [inline src/ray/raylet/node_manager.cc]: Comes after `node_manager.h`

**Comment by robertnishihara** [comment]
@ericl this looks incredibly useful.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**